timemachine/exchange/exchange_mover.py

```diff
@@ -87,108 +95,385 @@ def get_initial_state(water_pdb, mol, ff, seed, nb_cutoff):
+            )
+
+            assert len(old_water_water_ixn_nrgs) == self.num_waters - 1
 
-        self.U_fn = U_fn
+            old_water_water_ixn_nrgs_full = jnp.insert(old_water_water_ixn_nrgs, water_idx, 0)
+            new_conf = conf.copy()
+            new_conf = new_conf.at[a_idxs].set(new_pos)
+            new_water_ixn_nrgs = jnp.sum(self.beta * U_fn_unsummed(new_conf, box, a_idxs, b_idxs), axis=0)
+            new_water_water_ixn_nrgs = jnp.sum(
+                new_water_ixn_nrgs[: (self.num_waters - 1) * 3].reshape((self.num_waters - 1), 3), axis=1
+            )
+
+            new_water_water_ixn_nrgs_full = jnp.insert(new_water_water_ixn_nrgs, water_idx, 0)
+            final_weights = initial_weights - old_water_water_ixn_nrgs_full + new_water_water_ixn_nrgs_full
+            final_weights = final_weights.at[water_idx].set(jnp.sum(new_water_ixn_nrgs))
+
+            # (ytz): sanity check to ensure we're doing incremental log weights correctly
+            # note that jax 64bit needs to be enabled first.
+            # ref_final_weights = self.batch_log_weights(new_conf, box)
+            # np.testing.assert_almost_equal(np.array(final_weights), np.array(ref_final_weights))
+
+            return final_weights, new_conf
+
+        self.batch_log_weights_incremental = batch_log_weights_incremental
+        self.batch_log_weights = batch_log_weights
 
     def propose(self, x: CoordsVelBox) -> Tuple[CoordsVelBox, float]:
         coords = x.coords
         box = x.box
-        n_atoms = len(coords)
-        chosen_water = np.random.randint(self.num_waters)
-        chosen_water_atoms = self.water_idxs[chosen_water]
-
-        # compute delta_U of deletion
-        a_idxs = chosen_water_atoms
-        b_idxs = np.delete(np.arange(n_atoms), a_idxs)
+        log_weights_before = self.batch_log_weights(coords, box)
+        log_probs_before = log_weights_before - logsumexp(log_weights_before)
+        probs_before = np.exp(log_probs_before)
+        chosen_water = np.random.choice(np.arange(self.num_waters), p=probs_before)
+        chosen_water_atoms = self.water_idxs_np[chosen_water]
 
         # compute delta_U of insertion
         trial_chosen_coords = coords[chosen_water_atoms]
         trial_translation = np.diag(box) * np.random.rand(3)
         # tbd - what should we do  with velocities?
 
-        # this has a higher acceptance probability than if we allowed for rotations
-        # (probably because we have a much higher chance of a useless move)
-        moved_coords = randomly_translate(trial_chosen_coords, trial_translation)
-        trial_coords = coords.copy()  # can optimize this later if needed
-        trial_coords[chosen_water_atoms] = moved_coords
+        moved_coords = randomly_rotate_and_translate(trial_chosen_coords, trial_translation)
+        # moved_coords = randomly_translate(trial_chosen_coords, trial_translation)
+
+        # optimized version using double transposition
```

Comment:
doc request: add description of "the transposition trick" and "double transposition" in PR or code?

Reply:
I'll try and clarify this - TL;DR allows for computation of the denominator (comprised of summed exponentials) to be done significantly faster by computing only 3 rows and transposing it.